### ISSUE TITLE
Question about network updating

### ISSUE THREAD

**Issue by zen-d** [OpenDriveLab/UniAD#25]
@YTEP-ZHI Hello, thanks for your great work. When I set https://github.com/OpenDriveLab/UniAD/blob/main/projects/mmdet3d_plugin/uniad/detectors/uniad_track.py#L545-L546 to
```python
prev_img, prev_img_metas = None, None
```
I find that `memory_bank` and `query_interact` do not receive gradients. It is a bit hard for me to understand, could you please explain that? What confuses me more is that MUTR3D does not use temporal feature fusion and could run without such problems.

**Comment by YTEP-ZHI** [comment]
Hi, @zen-d. Thanks for your question. We will check on this soon. By the way, do they(mem_bank and query_interact) receive gradients and function normally when you unset this modification?

**Comment by YTEP-ZHI** [comment]
Let me explain the error in detail. The error of "not receiving gradients" is correlated with the setting: ` find_unused_parameters=False`. It's the default setting of a pytorch program, designed to warn programmers when some modules expected to receive gradients(`modules.requires_grad=True`)  are actually not producing any loss or not receiving gradients after the forward pass. But if you change the ` find_unused_parameters` to `True` as we did, the error will not occur.

So where did those unused parameters come from in UniAD? 
1. To save GPU memory, we do not update the image backbone and image neck in both two training stages of UniAD. We implement the `freeze backbone/neck` operation in the forward pass by forcing them not to produce any gradient here: https://github.com/OpenDriveLab/UniAD/blob/3643c68f8305d1531057ef973df06c45a5de229b/projects/mmdet3d_plugin/uniad/detectors/uniad_track.py#L176-L177
So if you set `find_unused_parameters=False`, the stop-gradient operation in image backbone and neck will trigger the forementioned error. A better option to freeze modules should be using `modules.requires_grad=True` when initialize modules instead of what we did in the forward pass.

3. UniAD aggregates multi-frame BEVs to enrich the representation of BEV in current timestep. However, in order to save GPU memory, we do not calculate any gradients when generating historical BEVs. It's also implemented in the forward pass similar to the stop-grad in backbone/neck. 
https://github.com/OpenDriveLab/UniAD/blob/3643c68f8305d1531057ef973df06c45a5de229b/projects/mmdet3d_plugin/uniad/detectors/uniad_track.py#L330-L341
It will also potentially trigger the error you mentioned when setting ` find_unused_parameters=False`.

**Comment by YTEP-ZHI** [comment]
Yes, it's inherited from BEVFormer. It's strange that a lot of people have used this repo to train models, without encoutering the gradient problem. I'll check on this. If you have any updated information, please report them in this thread. Thanks.

**Comment by YTEP-ZHI** [comment]
@zen-d 
Thanks for much for providing valuable information. I'm checking on the reproduction issue and gradient issue, and will notice you once it's fixed.

**Comment by YTEP-ZHI** [comment]
Hi, @zen-d. The gradient problem, that the `memory_bank` (sometimes) doesn't receive gradient, is triggered when `memory_bank` is empty (in the beginning of track training or starting at a new scenario). When it's empty (`len(embed) ==0`), the interactions in `memory_bank` will not be performed. Hence, the neural branches in `memory_bank` will not be forwarded and updated, which causes this issue.
https://github.com/OpenDriveLab/UniAD/blob/86aca41b1fb149ef2b69f0e8a967cd2c29d1cf48/projects/mmdet3d_plugin/uniad/dense_heads/track_head_plugin/modules.py#L64

However, I think it's acceptable to set `find_unused_parameters=True` to bypass this issue.

Moreover, our trackformer is modified from MOTR. You can find that they also set `find_unused_parameters=True` in their training codes.
https://github.com/OpenDriveLab/UniAD/blob/86aca41b1fb149ef2b69f0e8a967cd2c29d1cf48/projects/mmdet3d_plugin/uniad/dense_heads/track_head_plugin/modules.py#LL64C1-L64C1

**Comment by YTEP-ZHI** [comment]
> @YTEP-ZHI
> 
> 1. I think the cfg setting in [this line](https://github.com/OpenDriveLab/UniAD/blob/main/projects/configs/stage1_track_map/base_track_map.py#L569) somehow bypasses the gradient problem (just without an error explicitly printed out), that is why "a lot of people do not encounter that". However, it surfaces now when I set `find_unused_parameters=False`.
> 2. Based on my latest ablation experiments stated in this comment [Question about network updating #25 (comment)](https://github.com/OpenDriveLab/UniAD/issues/25#issuecomment-1555875001), I might conjecture it is not a trivial problem related to frozen network parts.
> 3. Accordingly, I also fall into reproduction issue [Could you share the training log of stage one? #21](https://github.com/OpenDriveLab/UniAD/issues/21) which is still pending. Could it be somehow related the gradient issue we discussed now?
> 
> Overall, look forward to your in-depth checked results and potential fix soon. Thanks.

The third issue is resolved as mentioned here: https://github.com/OpenDriveLab/UniAD/issues/21#issuecomment-1586549953, and the performance of stage1 model could reproduced when trained from scratch. It's actually nothinig to do with the gradient issue. Still, thanks for your feedback.

**Comment by YTEP-ZHI** [comment]
I'm closing this issue as it's resolved, feel free to reopen it if needed @zen-d.
